src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs

```diff
@@ -165,23 +165,27 @@ public override bool IsInvalid
             }
         }
 
-        // Specialized Open that returns the file length and permissions of the opened file.
+        // Specialized Open that returns the FileStatus of the opened file.
         // This information is retrieved from the 'stat' syscall that must be performed to ensure the path is not a directory.
-        internal static SafeFileHandle OpenReadOnly(string fullPath, FileOptions options, out long fileLength, out UnixFileMode filePermissions)
+        internal static SafeFileHandle OpenReadOnly(string fullPath, FileOptions options, out Interop.Sys.FileStatus status)
         {
-            SafeFileHandle handle = Open(fullPath, FileMode.Open, FileAccess.Read, FileShare.Read, options, preallocationSize: 0, DefaultCreateMode, out fileLength, out filePermissions, null);
-            Debug.Assert(fileLength >= 0);
+            SafeFileHandle handle = Open(fullPath, FileMode.Open, FileAccess.Read, FileShare.Read, options, preallocationSize: 0, DefaultCreateMode, out status, out bool isReadOnly);
+            Debug.Assert(isReadOnly);
             return handle;
         }
 
+        // Used by callers of OpenReadOnly.
+        internal static UnixFileMode GetFileMode(Interop.Sys.FileStatus status) =>
+            ((UnixFileMode)status.Mode) & PermissionMask;
+
         internal static SafeFileHandle Open(string fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize, UnixFileMode? unixCreateMode = null,
                                             Func<Interop.ErrorInfo, Interop.Sys.OpenFlags, string, Exception?>? createOpenException = null)
         {
             return Open(fullPath, mode, access, share, options, preallocationSize, unixCreateMode ?? DefaultCreateMode, out _, out _, createOpenException);
         }
 
         private static SafeFileHandle Open(string fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize, UnixFileMode openPermissions,
-                                            out long fileLength, out UnixFileMode filePermissions,
+                                            out Interop.Sys.FileStatus status, out bool isReadOnly,
```

Comment:
You've added `isReadOnly` as a way to implement the `Debug.Assert(fileLength >= 0);` check that ensures the value is initialized. The name `isReadOnly` doesn't reflect that. You can remove the parameter completely.